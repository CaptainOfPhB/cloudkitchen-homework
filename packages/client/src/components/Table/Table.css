.table-component {
  color: #333;
  width: 100%;
  font-size: 14px;
  border-spacing: 0;
  border-collapse: collapse;
}

.table-header {
  padding: 16px 8px;
  text-align: left;
}

.table-column-title {
  font-weight: bold;
  padding: 20px 8px;
  background-color: #f5f5f5;
  border-bottom: 1px solid #e5e5e5;
}

.table-row {
  display: table-row;
  vertical-align: inherit;
  border-color: inherit;
  border-bottom: 1px solid #eeeeee;
}

.table-empty-content {
  height: 200px;
  font-weight: 400;
  color: rgba(0, 0, 0, .25);
}

.table-cell {
  text-align: left;
  box-sizing: border-box;
  padding: 10px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}