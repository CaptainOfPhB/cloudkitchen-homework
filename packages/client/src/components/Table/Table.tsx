import './Table.css';
import type { CSSProperties, Key, ReactNode } from 'react';

interface Column<T> {
  dataIndex?: keyof T;
  title: ReactNode;
  style?: CSSProperties;
  render?: (value: any, record: T) => ReactNode;
}

interface TableProps<T> {
  dataSource: T[];
  columns: Column<T>[];
  header?: ReactNode;
  rowKey: keyof T;
}

function Table<T>(props: TableProps<T>) {
  return (
    <table className='table-component'>
      {props.header && (
        <thead>
        <tr>
          <th colSpan={props.columns.length} className='table-header'>
            {props.header}
          </th>
        </tr>
        </thead>
      )}
      <tbody>
      <tr className='table-column-title'>
        {props.columns.map(column => {
          return (
            <th
              className='table-cell'
              style={column.style}
              key={column.dataIndex as string}
            >
              {column.title}
            </th>
          );
        })}
      </tr>
      {props.dataSource.length
        ? props.dataSource.map(each => {
          return (
            <tr key={each[props.rowKey] as unknown as Key} className='table-row'>
              {props.columns.map(column => {
                const value = each[column.dataIndex as keyof T];
                return (
                  <td
                    className='table-cell'
                    style={column.style}
                    key={column.dataIndex as string}
                  >
                    <>{column.render ? column.render(value, each) : value}</>
                  </td>
                );
              })}
            </tr>
          );
        })
        : (
          <tr className='table-row'>
            <th rowSpan={5} colSpan={props.columns.length} className='table-empty-content'>
              No data
            </th>
          </tr>
        )}
      </tbody>
    </table>
  );
}

export default Table;