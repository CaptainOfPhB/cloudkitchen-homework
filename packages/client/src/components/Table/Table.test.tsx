import Table, { Column } from './Table';
import { render, screen } from '@testing-library/react';

interface Data {
  id?: number;
  name?: string;
  age?: number;
  gender?: string;
}

it('should render table on screen', () => {
  render(<Table dataSource={[]} columns={[]} rowKey='' />);
  const table = screen.getByTestId('table-component');
  expect(table).toMatchSnapshot();
});

it('should render \'No Data\' when list is empty', () => {
  render(<Table dataSource={[]} columns={[]} rowKey='' />);
  expect(screen.getByText('No Data')).toMatchSnapshot();
  expect(screen.getByText('No Data')).toBeInTheDocument();
});

it('should render table header', function () {
  const header = 'hello world';
  render(<Table dataSource={[]} columns={[]} rowKey='' header={header} />);
  expect(screen.getByText(header)).toBeInTheDocument();
});

it('should render columns and dataSource', function () {
  const columns: Column<Data>[] = [
    { title: 'id', dataIndex: 'id' },
    { title: 'name', dataIndex: 'name' },
  ];
  const dataSource = [{ id: 1, name: 'John' }, { id: 2, name: 'Jane' }];
  render(<Table dataSource={dataSource} columns={columns} rowKey='id' />);
  expect(screen.getByTestId('table-component')).toMatchSnapshot();
});

it('should call render callback with different params according to key and dataIndex', function () {
  const renderWhenKeyExist = jest.fn();
  const renderWhenDataIndexExist = jest.fn();
  const columns: Column<Data>[] = [
    { title: 'id', key: 'id', render: renderWhenKeyExist },
    { title: 'name', dataIndex: 'name', render: renderWhenDataIndexExist },
    { title: 'age', dataIndex: 'age' },
    { title: 'gender', key: 'gender' },
  ];
  const dataSource = [{ id: 1, name: 'John' }, { id: 2, name: 'Jane' }];
  render(<Table dataSource={dataSource} columns={columns} rowKey='id' />);
  expect(renderWhenDataIndexExist).toHaveBeenCalledTimes(2);
  expect(renderWhenKeyExist).toHaveBeenCalledTimes(2);
  expect(renderWhenDataIndexExist).toBeCalledWith(dataSource[0].name, dataSource[0], 0);
  expect(renderWhenKeyExist).toBeCalledWith(dataSource[0], 0);
});
