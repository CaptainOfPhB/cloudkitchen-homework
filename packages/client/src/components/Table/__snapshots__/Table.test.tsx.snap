// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 'No Data' when list is empty 1`] = `
<th
  class="table-empty-content"
  colspan="0"
  rowspan="5"
>
  No Data
</th>
`;

exports[`should render columns and dataSource 1`] = `
<table
  class="table-component"
  data-testid="table-component"
>
  <tbody>
    <tr
      class="table-column-title"
    >
      <th
        class="table-cell"
      >
        id
      </th>
      <th
        class="table-cell"
      >
        name
      </th>
    </tr>
    <tr
      class="table-row"
    >
      <td
        class="table-cell"
      >
        1
      </td>
      <td
        class="table-cell"
      >
        John
      </td>
    </tr>
    <tr
      class="table-row"
    >
      <td
        class="table-cell"
      >
        2
      </td>
      <td
        class="table-cell"
      >
        Jane
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render table on screen 1`] = `
<table
  class="table-component"
  data-testid="table-component"
>
  <tbody>
    <tr
      class="table-column-title"
    />
    <tr
      class="table-row"
    >
      <th
        class="table-empty-content"
        colspan="0"
        rowspan="5"
      >
        No Data
      </th>
    </tr>
  </tbody>
</table>
`;
